### ISSUE TITLE
[hmac,dv] Run UNR analysis

### ISSUE THREAD

**Issue by gdessouky** [lowRISC/opentitan#23683]
### Description

Once coverage reaches its maximum possible, run UNR anaylsis.

@martin-velay @moidx @vogelpi

**Comment by martin-velay** [comment]
I ran a regression today, but not with the master's head. I used this commit 7f47f7564ed63efee7b29022d5eb05441114024c, it takes several recent changes which are not yet merged.

Here are the results:
![image](https://github.com/lowRISC/opentitan/assets/165653139/db3d477c-f394-42fb-a211-3cfa4bd0f73e)

I think we should dig to understand why the FSM code cov is so weak. Now I'll have analyze the fcov.

**Comment by martin-velay** [comment]
I have done some analysis and I have some inputs to share about the coverage. I am here referring to [this regression](https://reports.opentitan.org/hw/ip/hmac/dv/2024.07.03_01.44.11/report.html):

1. prim_sha2_pad FSM::st_d from this instance `tb.dut.u_prim_sha2_512.gen_multimode_logic.u_prim_sha2_multimode.u_pad`.
Here are the uncovered transitions:
![image](https://github.com/lowRISC/opentitan/assets/165653139/01ab320c-3f14-47ac-984f-8b1f1a7a349e)
Which are all referring to the same RTL line: [prim_sha2_pad.sv @line341](https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/prim/rtl/prim_sha2_pad.sv;l=341).
To cover this we can try to disable the SHA core randomly while a message is being written, I am currently trying this locally.
2. prim_sha2 FSM::sha_st_q from this instance `tb.dut.u_prim_sha2_512.gen_multimode_logic.u_prim_sha2_multimode`.
Here is the uncovered transition:
![image](https://github.com/lowRISC/opentitan/assets/165653139/248f863f-8ca7-42d4-bcd8-938e02927c17)
RTL line: [prim_sha2.sv @line432](https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/prim/rtl/prim_sha2.sv;l=432).
@gdessouky any idea how to trigger this?
3. Initially I though that some other FSM were not recognized but it seems that all the others are classified correctly and fully covered @andreaskurth and @vogelpi

**Comment by gdessouky** [comment]
From a regression run (at commit https://github.com/lowRISC/opentitan/commit/e9ae10fb426b7a48caf2ec9bdb4e5c723270fe5e) with the UNR exclusions generated (see https://github.com/lowRISC/opentitan/pull/23916), the coverage increases to

### [Coverage Dashboard](/workspaces/repo/scratch/hmac_unr/hmac-sim-vcs/cov_report/dashboard.html)

|  SCORE  |  LINE   |  COND   |  TOGGLE  |   FSM   |  BRANCH  |  ASSERT  |  GROUP  |
|:-------:|:-------:|:-------:|:--------:|:-------:|:--------:|:--------:|:-------:|
| 96.87 % | 97.15 % | 97.22 % | 100.00 % | 89.47 % | 98.43 %  | 99.49 %  | 96.35 % |

So there's a good boost there. Looking into the FSM coverage:
![image](https://github.com/lowRISC/opentitan/assets/5173875/27b40a25-232c-47c0-b2fb-1eb9e11b1344)
this transition is now excluded, so we can't trigger this. At a much later phase, beyond RTL freeze, we could trim this out I guess. I recall now seeing that one in the UNR list some time ago still.

These transitions though remain unexcluded.
> 
> 1. prim_sha2_pad FSM::st_d from this instance `tb.dut.u_prim_sha2_512.gen_multimode_logic.u_prim_sha2_multimode.u_pad`.
>    Here are the uncovered transitions:
>    ![image](https://private-user-images.githubusercontent.com/165653139/345827883-01ab320c-3f14-47ac-984f-8b1f1a7a349e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxNTU2MDYsIm5iZiI6MTcyMDE1NTMwNiwicGF0aCI6Ii8xNjU2NTMxMzkvMzQ1ODI3ODgzLTAxYWIzMjBjLTNmMTQtNDdhYy05ODRmLThiMWYxYTdhMzQ5ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwNVQwNDU1MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MDA3MTA0NzNhMWUyMDY5OGQ0NGY2ZjIyNWY0N2QyMjNlYjU2YTk2Y2Y0ZTgzOGU1ZWEyZDc0YWRkNzk2YWU5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QHfFcwZRJTB66Gjbu_ZrjoqtI7vzkvBPU9Ys30EBOQc)

These may be logically impossible to trigger because of how the padding logically works? I would look into it a bit, but surprisingly I see them excluded in the older UNR exclusions generated here that's on master https://github.com/lowRISC/opentitan/blob/master/hw/ip/hmac/dv/cov/hmac_unr_excl.el#L85-L97 from Jan 10, and I would have expected to see them excluded in this last UNR run! Perhaps something changed in the RTL around it that prevented that, I'll look into it a bit.  But once either this is covered or excluded (likely what should happen), the FSM coverage would probably up to ~95%.

**Comment by gdessouky** [comment]
@martin-velay as as you point out, this line here https://github.com/lowRISC/opentitan/blob/master/hw/ip/prim/rtl/prim_sha2_pad.sv#L341 was added prior to the previous UNR run, and is why these transitions are not captured as uncovered in the recent UNR run. I guess the use case here is if a new hashing operation is triggered with start/continue to interrupt an ongoing one, which makes sense. Except that we actually block that by triggering an error code https://github.com/lowRISC/opentitan/blob/master/hw/ip/hmac/rtl/hmac.sv#L801 if I understand correctly, so it will not flow into the padding FSM in these states to trigger a transition to `StFifoReceive`.

We can always manually exclude the transitions, or see if we want to remove that line?

**Comment by vogelpi** [comment]
After discussing with @andreaskurth , we have come to the conclusion that these transitions are actually due to an RTL bug. We should not invest time to try to cover them and instead remove the transition in a future release. I'll file an issue. The good thing is that the RTL bug isn't critical at all.

@gdessouky would you mind writing a manual exclusion for these transitions please?

**Comment by gdessouky** [comment]
Thanks, I agree @vogelpi, this isn't intended behavior and exercising it is essentially trying to achieve coverage for a bug/unrequired RTL line. I'll add the manual exclusions in a PR shortly

**Comment by andreaskurth** [comment]
Thanks @vogelpi and @gdessouky. I agree and just wanted to note that this RTL bug will not manifest itself as long as HMAC is used according to specification (programming model).

**Comment by andreaskurth** [comment]
@gdessouky Can we close this once #23948 is merged?
